lib/KettleMiddleware-Multer.js

```diff
@@ -0,0 +1,229 @@
+    invokers: {
+        multerFileFilter: {
+            funcName: "kettle.middleware.multer.filter.mimeType.filterFunc",
+            args: ["{arguments}.0", "{arguments}.1", "{that}.options.acceptedMimeTypes", "{arguments}.2"]
+        }
+    }
+});
+
+// Filters by mimeType based on a list of accepted mime types
+kettle.middleware.multer.filter.mimeType.filterFunc = function (req, file, acceptedMimeTypes, cb) {
+    var isAcceptableMimeType = fluid.contains(acceptedMimeTypes, file.mimetype);
+    cb(null, isAcceptableMimeType);
+};
+
+// Abstract grade for storage
+fluid.defaults("kettle.middleware.multer.storage", {
+    gradeNames: ["fluid.component"],
+    // This must be one of the supported Multer
+    // storage functions (memoryStorage or diskStorage,
+    // unless using additional plug-ins for Multer);
+    // see the concrete grades below
+    invokers: {
+        multerStorage: {
+            funcName: "fluid.notImplemented"
+        }
+    }
+});
+
+// Storage grade using Multer's memoryStorage
+// storage function - uploaded files are only
+// stored in memory for the duration of the
+// request, although a handler or other
+// middleware can process them from
+// request.req.file or request.req.files
+fluid.defaults("kettle.middleware.multer.storage.memory", {
+    gradeNames: ["kettle.middleware.multer.storage"],
+    invokers: {
+        multerStorage: {
+            funcName: "kettle.npm.multer.memoryStorage",
+            args: []
+        }
+    }
+});
+
+// Storage grade using Multer's diskStorage
+// storage function, with destination and
+// filename as configurable options
+//
+// This grade should be adequate for most
+// common uses of disk storage - complex
+// uses should derive from this grade and
+// replace the destination and filename
+// invokers as necessary to return
+// appropriate functions in Multer's
+// expected style
+fluid.defaults("kettle.middleware.multer.storage.disk", {
+    gradeNames: ["kettle.middleware.multer.storage"],
+    // "destination" can be a plain string or a string using fluid.stringTemplate
+    // syntax - the template will receive the "file information" key-values
+    // described at https://github.com/expressjs/multer#file-information
+    destination: "./tests/data/uploads",
+    // "filename" should be a string using fluid.stringTemplate
+    // syntax - the template will receive the "file information" key-values
+    // described at https://github.com/expressjs/multer#file-information
+    //
+    // Note that simply using a plain string will result in every uploaded
+    // file overwriting the previous one - this is unlikely to be desired
+    // behaviour
+    filename: "%originalname",
+    members: {
+        // TODO: is this the best way to be able to supply an
+        // object to an invoker where the object uses IoC references
+        // for the values of its keys?
+        diskStorageConfigurationObject: {
+            destination: "{that}.destination",
```

Comment:
Some kind of big problem here since both of these resolve to `undefined` since they should be referenced as {that}.options.destination. Suggests that the actual values here are never tested, and only the values supplied to the invokers are important. In that case it should probably be removed. If it is retained, it's ok for it to be encoded within "options" as with its source values, but somehow it should receive dedicated tests if it is kept.

Reply:
This is confusing because of bad naming (which I'll correct) - `{that}.options.destination` (a configuration value) and `{that}.destination` (an invoker) are both valid and separate, but have identical names. I'll change the invokers to `destinationImpl` and `filenameImpl` or similar so this is clearer.

Multer's own `diskStorage` function expects two functions in a key-value structure like so:

```
{
  destination: function(req, file, cb)...,
  filename: function(req, file, cb)...
}
```

So the option values are used as arguments to the invoker functions (which use argument-shifting to provide the three-value signatures Multer requires), and the invoker functions are referenced in the `diskStorageConfigurationObject`, which is passed into the actual Multer `diskStorage` function by the `multerStorage` invoker.

I couldn't find an invoker syntax that would allow me to pass a configuration object with IoC-resolved values directly, hence the storage

I find it confusing explaining this and I wrote the wrapping code, but it's how Multer's internals work - the implementation should hide this from an end user and allow most use cases to simply set `filename` and `destination` to string values with `fluid.stringTemplate` syntax.

Reply:
Hi Alan, sorry to misread the code, but I guess this does highlight that it could be confusing. Perhaps "destinationResolver" or "destinationCallback" would be clearer in terms of conveying the different roles of the various incarnations of these values.

I'm not quite sure what you mean by "an invoker syntax that would allow me to pass an object with IoC-resolved values directly" - but perhaps you just mean something like this?

````
        multerStorage: {
            funcName: "kettle.npm.multer.diskStorage",
            "args": [{
                destination: "{that}.destinationResolver",
                filename: "{that}.filenameResolver"
             }]
        }
````

This will let you do without the indirection onto `diskStorageConfigurationObject`. If this wasn't what you were expecting, perhaps you could list a couple of the things you tried that didn't seem to work?